nixos/modules/services/networking/rosenpass.nix

```diff
@@ -0,0 +1,178 @@
+              options = {
+                publicKey = mkOption {
+                  type = str;
+                  description = "WireGuard public key corresponding to the remote Rosenpass peer.";
+                };
+              };
+            };
+            description = "WireGuard configuration for this peer.";
+          };
+        };
+      };
+    in
+      mkOption {
+        type = listOf peer;
+        description = "List of peers to exchange keys with.";
+        default = [];
+      };
+
+    extraConfig = mkOption {
+      type = attrs;
+      description = ''
+        Extra configuration to be merged with the generated Rosenpass configuration file.
+      '';
+      default = {};
+    };
+  };
+
+  config = mkIf cfg.enable {
+    warnings = let
+      netdevsList = attrValues config.systemd.network.netdevs;
+      publicKeyInNetdevs = peer: any (netdev: any (publicKeyInWireguardPeers peer) netdev.wireguardPeers) netdevsList;
+      publicKeyInWireguardPeers = peer: x: x.wireguardPeerConfig ? PublicKey && x.wireguardPeerConfig.PublicKey == peer.wireguard.publicKey;
+
+      # NOTE: In the message below, we tried to refer to
+      #   options.systemd.network.netdevs."<name>".wireguardPeers.*.PublicKey
+      # directly, but don't know how to traverse "<name>" and * in this path.
+      warningMsg = peer: "It appears that you have configured a Rosenpass peer with the Wireguard public key '${peer.wireguard.publicKey}' but there is no corresponding Wireguard peer configuration in any of `${options.systemd.network.netdevs}.\"<name>\".wireguardPeers.*.PublicKey`. While this may work as expected, such a scenario is unusual. Please double-check your configuration.";
+    in
+      concatMap (peer: optional (!publicKeyInNetdevs peer) (warningMsg peer)) cfg.peers;
+
+    environment.systemPackages = [cfg.package pkgs.wireguard-tools];
+
+    users.users."${cfg.user}" = {
+      isSystemUser = true;
+      createHome = false;
+      group = cfg.group;
+    };
+
+    users.groups."${cfg.group}" = {};
+
+    # NOTE: It would be possible to use systemd credentials for pqsk.
+    # <https://systemd.io/CREDENTIALS/>
+    systemd.services.rosenpass = let
+      generatePeerConfig = {
+        publicKeyFile,
+        endpoint,
+        device,
+        wireguard,
+      }:
+        {
+          inherit device;
+          public_key = publicKeyFile;
+          peer = wireguard.publicKey;
+          extra_params = [];
+        }
+        // (optionalAttrs (endpoint != null) {inherit endpoint;});
+
+      generateConfig = {
+        publicKeyFile,
+        secretKeyFile,
+        listen,
+        verbosity,
+        peers,
+        ...
+      }: {
+        inherit listen verbosity;
+        public_key = publicKeyFile;
+        secret_key = secretKeyFile;
+        peers = map generatePeerConfig peers;
+      };
+      toml = pkgs.formats.toml {};
+      configFile = toml.generate "config.toml" (recursiveUpdate (generateConfig cfg) cfg.extraConfig);
```

Comment:
With a few adjustments, you can implement [RFC 42](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md) more fully:

- Instead of standalone config options and `extraConfig`, put them all in a `settings` option defined as a [freeform module](https://nixos.org/manual/nixos/stable/index.html#sec-freeform-modules).

- Keep options like `public_key` named as they are in the config file. You can still define them as explicit options inside the `settings` freeform module to give them a description and validate against their type.

- Pass the whole `settings` value to `toml.generate`.